train.py

```diff
@@ -0,0 +1,176 @@
+import os
+
+import hydra
+import torch
+import wandb
+import torchmetrics
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from omegaconf import OmegaConf
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from models.unet import Unet
+from ecg_segmentation_dataset import ECGDataset
+
+
+def intersection_over_union(input: torch.Tensor, target: torch.Tensor, threshold: float = 0.5):
+    # input and target are probabilities of belonging to certain class
+    pred = input > threshold
+    mask = target > threshold
+
+    intersection = (pred & mask).sum()
+    union = (pred | mask).sum()
+
+    # if union is all zeros, it means signal was zeros and model predicted it well, so return 1
+    iou = intersection / union if union > 0 else 1
+
+    return iou
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+def preprocess_dataset(dataset_name: str, batch_size: int, num_workers: int):
+    train_ds = ECGDataset(dataset_name, split="train")
+    val_ds = ECGDataset(dataset_name, split="validation")
+    test_ds = ECGDataset(dataset_name, split="test")
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
+    val_dataloader = DataLoader(val_ds, batch_size=batch_size, num_workers=num_workers)
+    test_dataloader = DataLoader(test_ds, batch_size=batch_size, num_workers=num_workers)
+
+    return train_dataloader, val_dataloader, test_dataloader
+
+
+def save_onnx_model(model: nn.Module, path: str):
+    device = next(model.parameters()).device
+    # example input for model
+    dummy_input = torch.randn((1, 2, 1000), requires_grad=True, device=device)
+
+    # exporting onnx model
+    torch.onnx.export(model, dummy_input, path)
+
+
+def save_checkpoint(model: nn.Module, optimizer: optim.Optimizer, save_path: str):
+        # saving models
+        torch.save({"model": model.state_dict(), "optimizer": optimizer.state_dict()}, save_path)
+
+def step(
+        model: nn.Module,
+        batch: dict[str, torch.Tensor, torch.Tensor],
+        device: torch.device
+    ):
+    # extract signal and mask from batch
+        signal = batch["signal"].to(device)
+        mask = batch["mask"].to(device)
```

Comment:
```suggestion
def save_checkpoint(model: nn.Module, optimizer: optim.Optimizer, save_path: str):
      # saving models
      torch.save({"model": model.state_dict(), "optimizer": optimizer.state_dict()}, save_path)

def step(
    model: nn.Module,
    batch: dict[str, torch.Tensor, torch.Tensor],
    device: torch.device
):
    # extract signal and mask from batch
    signal = batch["signal"].to(device)
    mask = batch["mask"].to(device)
```

Both those functions have indent issues - do you now how did this get pass pre commit hooks? 🤔